File: app/repository/TournamentRepository.java
package repository;

import models.Tournament;
import play.db.ebean.EbeanConfig;

import javax.inject.Inject;
import java.util.Optional;
import java.util.concurrent.CompletionStage;


/**
 * A repository that executes database operations in a different
 * execution context.
 */
public class TournamentRepository extends BaseRepository<Tournament> {

    @Inject
    public TournamentRepository(EbeanConfig ebeanConfig, DatabaseExecutionContext executionContext) {
        super(ebeanConfig, executionContext);
    }

    @Override
    protected Class<Tournament> getModelClass() {
        return Tournament.class;
    }

    public CompletionStage<Optional<Tournament>> view(Long id) {
        return get(id, query -> query
                .fetch("challenge", "id,name")
                .fetch("entries", "id,rank")
                .fetch("entries.version", "id,name")
                .fetch("games", "id,startTime,endTime"));
    }
}

File: app/repository/TournamentRepository.scala
package repository

import javax.inject._
import models.Tournament
import play.db.ebean.EbeanConfig

import scala.concurrent.Future

class TournamentRepository @Inject()(val ebeanConfig: EbeanConfig, val executionContext: DatabaseExecutionContext) extends BaseRepository[Tournament] {


  val modelClass: Class[Tournament] = classOf[Tournament]

  def view(id: Long): Future[Option[Tournament]] =
    getOneWhere(id) {
      query
        .fetch("challenge", "id,name,status")
        .fetch("challenge.owner", "id,")
        .fetch("challenge.activeTournament", "id")
        .fetch("entries", "id,rank")
        .fetch("entries.version", "id,name")
        .fetch("games", "id,startTime,endTime")
    }

  def all(challengeId: Long): Future[List[Tournament]] =
    getList {
      query
        .where().eq("challenge.id", challengeId)
        .orderBy("createdAt")
        .select("name,createdAt")
        .fetch("owner", "id,name")
    }

  def count(challengeId: Long): Future[Int] =
    execute {
      query.where().eq("challenge.id", challengeId).findCount()
    }

}
